### ISSUE TITLE
Negative valued scores from resnet and ft-transformer models

### ISSUE THREAD

**Issue by classopen24** [yandex-research/rtdl#8]
Hello - I have run into a puzzling issue with getting negative scores saved in the p_train/val/test files for the resnet and fttransformer models with my datasets. Given the structure of the models and from what I can tell in the model code, I don't see any obvious issues that would explain this. Would you have any suggestions for debugging or dealing with this issues? Appreciate any suggestions, thanks.

**Comment by classopen24** [comment]
Thanks again for your help. Can you confirm that, for binary class predictions, the conversion from model output logits to probabilities is performed in calculate metrics on line 33, by applying scipy.special.expit ?

**Comment by Yura52** [comment]
It depends. For catboost and xgboost, no transformation is applied; for mlp/resnet/..., the inverse of the logit function is applied; for ensembles, there is a separate script that handles this. Could you add some context and describe the issue, please?

**Comment by classopen24** [comment]
Thanks - we have been training NODE/FT-Transformer/ResNet models  for binary classification. In the context of binary classification with these models, does line 33 (scipy.special.expit) fully capture what needs to be done to convert model scores to probabilities?

**Comment by Yura52** [comment]
Let me start from scratch.

Let's consider two kinds of probabilities that you can obtain based on logits:

1. "raw probabilities": the result of `expit` (`softmax` for multiclass)
2. "calibrated probabilities": the result of calibration of logits or "raw probabilities". The following links can be helpful:
    - https://scikit-learn.org/stable/modules/calibration.html
    - https://www.cs.cornell.edu/~alexn/papers/calibration.icml05.crc.rev3.pdf
    - https://arxiv.org/abs/1706.04599

What is better for you depends on your use case (that is, how you are going to use probabilities), there is no "correct" answer.

Note that:
- You do not need probabilities to predict labels with a single model. In fact, this is exactly the case of `calculate_metrics`: the variable `probs` is completely unnecessary for making prediction and is presented for rather historical reasons. The prediction can be implemented using the raw logits:
    ```python
    # binclass
    labels = (logits > 0).astype(np.int64)

    # multiclass
    labels = logits.argmax(axis=1)    
    ```
- You need probabilities to predict labels with an ensemble of models (probabilities are used to calculate the average prediction, see bin/ensemble.py)

**Comment by classopen24** [comment]
This is very helpful and answers my question - thanks again.
